src/mlpack/methods/adaboost/adaboost_predict_proba_main.cpp

```diff
@@ -0,0 +1,101 @@
+/**
+ * @file methods/adaboost/adaboost_predict_proba_main.cpp
+ * @author Udit Saxena
+ *
+ * Implementation of the AdaBoost main program.
+ *
+ * @code
+ * @article{Schapire:1999:IBA:337859.337870,
+ *   author = {Schapire, Robert E. and Singer, Yoram},
+ *   title = {Improved Boosting Algorithms Using Confidence-rated Predictions},
+ *   journal = {Machine Learning},
+ *   issue_date = {Dec. 1999},
+ *   volume = {37},
+ *   number = {3},
+ *   month = dec,
+ *   year = {1999},
+ *   issn = {0885-6125},
+ *   pages = {297--336},
+ *   numpages = {40},
+ *   url = {http://dx.doi.org/10.1023/A:1007614523901},
+ *   doi = {10.1023/A:1007614523901},
+ *   acmid = {337870},
+ *   publisher = {Kluwer Academic Publishers},
+ *   address = {Hingham, MA, USA},
+ *   keywords = {boosting algorithms, decision trees, multiclass classification,
+ *   output coding}
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/util/io.hpp>
+
+#ifdef BINDING_NAME
+  #undef BINDING_NAME
+#endif
+#define BINDING_NAME adaboost_predict_proba
+
+#include <mlpack/core/util/mlpack_main.hpp>
+#include <mlpack/core/data/normalize_labels.hpp>
+#include "adaboost.hpp"
+#include "adaboost_model.hpp"
+
+using namespace mlpack;
+using namespace std;
+using namespace arma;
+using namespace mlpack::adaboost;
+using namespace mlpack::tree;
+using namespace mlpack::perceptron;
+using namespace mlpack::util;
+
+// Program Name.
+BINDING_USER_NAME("AdaBoost Probability Prediction");
+
+// Short description.
+BINDING_SHORT_DESC("Class probabilities from model.");
+
+// Long description.
+BINDING_LONG_DESC("");
+
+// Example.
+BINDING_EXAMPLE(
+    CALL_METHOD("model", "predict_proba", "test", "X_test"));
+
+// Classification options.
+PARAM_MATRIX_IN("test", "Test dataset.", "T");
+PARAM_MATRIX_OUT("probabilities", "Predicted class probabilities for each "
+    "point in the test set.", "p");
+
+// Loading/saving of a model.
+PARAM_MODEL_IN(AdaBoostModel, "input_model", "Input AdaBoost model.", "m");
+
+void BINDING_FUNCTION(util::Params& params, util::Timers& timers)
+{
+  // Check input parameters and issue warnings/errors as necessary.
+
+  RequireOnlyOnePassed(params, {"test"});
+  RequireOnlyOnePassed(params, {"input_model"});
```

Comment:
I think `RequireOnlyOnePassed` is not necessary here, since it checks that only one option in the given set of options was passed.  So, what this does is just ensure that `test` and `input_model` were passed---but another way you could handle that is just to use `PARAM_MATRIX_IN_REQ()` and `PARAM_MODEL_IN_REQ()` to achieve the same result.  (This applies to `adaboost_predict_main.cpp` too.)  :+1:

Reply:
I have a small doubt here. PrintInputProcessing output for a required matrix parameter is:

```cython
  # Detect if the parameter was passed; set if so.
  training_tuple = to_matrix(training, dtype=np.double, copy=p.Has('copy_all_inputs'))
  if len(training_tuple[0].shape) > 2:
    training_tuple[0].shape = (training_tuple[0].shape[0], 1)
  training_mat = arma_numpy.numpy_to_mat_d(training_tuple[0], training_tuple[1])
  SetParam[arma.Mat[double]](p, <const string> 'training', dereference(training_mat))
  p.SetPassed(<const string> 'training')
  del training_mat
```

and for a non-required parameter is:

```cython
  if training is not None:
    training_tuple = to_matrix(training, dtype=np.double, copy=p.Has('copy_all_inputs'))
    if len(training_tuple[0].shape) < 2:
      training_tuple[0].shape = (training_tuple[0].shape[0], 1)
    training_mat = arma_numpy.numpy_to_mat_d(training_tuple[0], training_tuple[1])
    SetParam[arma.Mat[double]](p, <const string> 'training', dereference(training_mat))
    p.SetPassed(<const string> 'training')
    del training_mat
```

apart from the first if condition (that just checks of the parameter is passed or not) shouldn't the entire code be the same?
why is there a ` > 2` for required and ` < 2` for a required parameter here?

Reply:
I think you are right!  This seems to be a bug.  The intention of that code, if I remember right, was to make it so that if a user passed a vector, we would automatically cast it to a matrix.  So, the correct one is `< 2`, not `> 2`.  Want to also make a change to fix this for the required parameter case?

Nice catch!

Reply:
changed [here](https://github.com/NippunSharma/mlpack/blob/e3df534edab5ec59374ab042f9bd4dfb26b58065/src/mlpack/bindings/python/print_input_processing.hpp#L356)